modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/programs/Genesis.scala

```diff
@@ -96,24 +102,44 @@ object Genesis {
       _ <- logger.info(s"Genesis binary ${binaryHash.show} and ${incrementalBinaryHash.show} accepted and sent to Global L0")
     } yield ()
 
-    def accept(
-      path: Path,
-      dataApplication: Option[BaseDataApplicationL0Service[F]] = None
-    )(implicit context: L0NodeContext[F]): F[Unit] = {
+    override def accept(dataApplication: Option[BaseDataApplicationL0Service[F]])(genesisPath: Path)(
+      implicit context: L0NodeContext[F]
+    ): F[Unit] = genesisLoader
+      .loadSignedGenesis(genesisPath)
+      .flatTap { genesis =>
+        dataApplication
+          .map(app => app.setCalculatedState(genesis.ordinal, app.genesis.calculated))
+          .getOrElse(false.pure[F])
+          .void
+      }
+      .flatMap(acceptSignedGenesis(dataApplication))
+
+    def create(dataApplication: Option[BaseDataApplicationL0Service[F]] = None)(
+      balancesPath: Path,
+      genesisOutputPath: Path,
+      keyPair: KeyPair
+    ): F[(Signed[CurrencySnapshot], Address)] = {
       def mkBalances =
         genesisLoader
-          .load(path)
+          .loadBalances(balancesPath)
           .map(_.map(a => (a.address, a.balance)).toMap)
 
       def mkDataApplicationPart =
         dataApplication.traverse(da => da.serializedOnChainGenesis.map(DataApplicationPart(_, List.empty, Hash.empty)))
 
-      (mkBalances, mkDataApplicationPart).mapN {
-        case (balances, dataApplicationPart) =>
-          CurrencySnapshot.mkGenesis(balances, dataApplicationPart)
-      }.flatTap { genesis =>
-        dataApplication.map(app => app.setCalculatedState(genesis.ordinal, app.genesis.calculated)).getOrElse(false.pure[F]).void
-      }.flatMap(accept(dataApplication))
+      def createGenesis = for {
+        balances <- mkBalances
+        dataApplicationPart <- mkDataApplicationPart
+        genesis = CurrencySnapshot.mkGenesis(balances, dataApplicationPart)
+        signedGenesis <- genesis.sign(keyPair)
+        signedBinary <- stateChannelSnapshotService.createGenesisBinary(signedGenesis)
+        identifier = signedBinary.value.toAddress
+      } yield (signedGenesis, identifier)
+
+      createGenesis.flatTap {
+        case (signedGenesis, identifier) =>
+          genesisLoader.write(signedGenesis, identifier, genesisOutputPath).void
+      }
```

Comment:
```suggestion
        _ <- genesisLoader.write(signedGenesis, identifier, genesisOutputPath)
      } yield (signedGenesis, identifier)
```
then you can discard `createGenesis` if you like and just have the for-comp.

Reply:
Sure. I had it extracted before for some reasons but implementation changed so I see that it can just single for-comp.